File: python-3/multiformats/multiformats/unsignedvarint.py
from typing import *

PRACTICAL_MAX_BYTES = 9
PRACTICAL_MAX = 2 ** (7 * PRACTICAL_MAX_BYTES) - 1

UnsignedVarint = NewType("UnsignedVarint", bytes)

def _encode(i: int) -> UnsignedVarint:
    assert(i >= 0)
    if i > 127:
        r = (bytes([(i % 128) + 128]) + encode (i // 128))
    else:
        r = bytes([i])
    return UnsignedVarint(r)

def encode(i: int) -> UnsignedVarint:
    assert(i <= PRACTICAL_MAX)
    return _encode(i)

def _decode(b: List[int]) -> int:
    match b:
        case []:
            raise Exception("Expected at least 1 byte")
        case [c, *t]:
            if c >= 128:
                assert(len(t) > 0)
                return (c - 128) + 128 * _decode(t)
            else:
                assert(len(t) == 0)
                return c

def decode(b: UnsignedVarint) -> int:
    assert(len(b) <= PRACTICAL_MAX_BYTES)
    return _decode(list(b))

File: kotlin/multiformats/lib/src/main/kotlin/multiformats/UnsignedVarInt.kt
package multiformats

import util.extensions.pow
import util.types.NonNegativeBigInteger

// NOTE: ULong (64-bits) is more than enough to meet current spec of maximum 9 bytes (7-bits)
data class UnsignedVarInt(val int: ULong) {
    companion object {
        const val PRACTICAL_MAX_BYTES: Int = 9
        val PRACTICAL_MAX: ULong = 128.toULong().pow(PRACTICAL_MAX_BYTES) - 1u

        fun digits(int: ULong): Int {
            require(int <= PRACTICAL_MAX)

            var exp = 0
            while ((128.toULong().pow(exp) - 1u) < int) exp += 1

            return exp
        }

        fun encode(int: ULong): ByteArray {
            require(int <= PRACTICAL_MAX)

            val digits = digits(int)
            return ByteArray(digits) { idx -> (
                int.div((128.toULong().pow(idx)))
                    .rem(128u) + if (idx == digits - 1) 0u else 128u
                ).toByte()
            }
        }

        @JvmStatic
        fun decode(bytes: ByteArray): ULong {
            require(bytes.isNotEmpty())
            require(bytes.size <= PRACTICAL_MAX_BYTES)
            require(bytes.dropLast(1).all { it.toUByte() >= 128u })
            require(bytes.last().toUByte() < 128u)

            return bytes.mapIndexed { idx, byte ->
                128.toULong().pow(idx) * byte.toUByte().mod(128u)
            }.reduce { a, b -> a + b }
        }
    }

    constructor(bytes: ByteArray) : this(decode(bytes))
    constructor(nnint: NonNegativeBigInteger) : this(nnint.int.longValueExact().toULong())

    init {
        require(int <= PRACTICAL_MAX)
    }

    val bytes: ByteArray by lazy { encode(int) }
}
